### PR TITLE
Gpl 773 Sample Sheet v10 support

### PR DESCRIPTION
**Pull request by BenTopping**
Coupled with Traction Service [PR](https://github.com/sanger/traction-service/issues/509)

**Changes proposed in this pull request:**

- Removed Sequencing Mode from Pacbio Well
- Removed Generate CCS Data
- Added Generate Hifi Reads fields to the Pacbio Well. The options provided for generate_hifi_reads is based of what the Runs System Name is. 
e.g. if System Name is "Sequel I" or "Sequel II", options = "**In SMRT Link**" and "Do Not Generate"
if System Name is "Sequel IIe", options are above aswell as "**On Instrument**" 
Bold above are the default values
- Added the default 2 to PreExtensionTime on a PacbioWell
- Added "Sequel IIe" to System Name options for a Pacbio Run
- Refactored PacbioRun building of a run so that it doesn't create 96 empty wells upfront. This makes it inline with what happens on update, where wells are then build when required.
- Fixed bug that meant if wells has been clicked on, they would build a well in the currentRun, and therefor would be sent to the service on create/update, which would then fail as well values were empty. (Related to the clean up suggestion below in TODO)
- Added CCS Analysis Output

Bug:
- Known bug is if the well already exists (in regards to the currentRun.plate.wells etc) then the 'buildWell' method isn't triggered on clicking the well, therefor the default isn't set. Suggestion is to somehow "clean up" the empty wells in the currentRun.

### CODE REVIEW

**Comment by harrietc52** [comment]
TODO:
- Add NPM

File: src/api/PacbioRun.js [hunk 1]
```diff
@@ -1,41 +1,19 @@
...
-const columns = () => {
-  return Array.from(Array(NO_OF_COLUMNS), (e, i) => String(i + 1))
-}
-
-const rows = () => {
-  return Array.from(Array(NO_OF_ROWS), (e, i) => String.fromCharCode(65 + i))
-}
-
-const buildWell = (row, column) => {
+const buildWell = (row, column, generateHiFiDefault='', preExtenstionTimeDefault='2') => {
```
Comment (stevieing) [inline src/api/PacbioRun.js:3]: you can turn this into an object literal which negates the need to have it as a function. I'm not sure that using generateHiFiDefault is the best name as it is an argument with a default. If you make it the same as the attribute then you can use the spread operator e.g.

```
const buildWell = (row, column, generate_hifi='', pre_extension_time='2') => ({
  row, 
  column,
  position: `${row}${column}`,
  movie_time: '',
  insert_size: '',
  on_plate_loading_concentration: '',
  sequencing_mode: '',
  generate_hifi,
  libraries: [],
  pre_extension_time
})
```

**Review by stevieing — APPROVED** [review]
All seems fine. You are right. A lot of refactoring is needed some of the files are getting quite big and there is a lot of duplication.

File: src/api/PacbioRun.js [hunk 2]
```diff
@@ -12,9 +13,10 @@ const buildWell = (row, column, generate_hifi='', pre_extension_time='2') => ({
     ccs_analysis_output: '',
     libraries: [],
     pre_extension_time
+    // pre_extension_time: preExtenstionTimeDefault,
```
Comment (stevieing) [inline src/api/PacbioRun.js]: There seem to be a few bits commented out. Don't know if they still need to be here?

File: src/components/pacbio/PacbioWellModal.vue [hunk 3]
```diff
@@ -56,7 +56,7 @@
           </b-form-select>
         </b-form-group>
 
-        <b-form-group v-if="showCCSAnalysis"
+        <!-- <b-form-group v-if="showCCSAnalysis"
```
Comment (stevieing) [inline src/components/pacbio/PacbioWellModal.vue]: More bits commented out.

File: src/components/pacbio/PacbioWellModal.vue [hunk 4]
```diff
@@ -139,7 +153,8 @@ export default {
         "Sequel II": ['In SMRT Link', 'Do Not Generate'],
         "Sequel IIe": ['In SMRT Link', 'Do Not Generate', 'On Instrument'],
       },
-      ccsAnalysisOptions: [{ text: 'CCS Analysis Output', value: "" }, 'Yes', 'No']
+      // ccsAnalysisOptions: [{ text: 'CCS Analysis Output', value: "" }, 'Yes', 'No']
```
Comment (stevieing) [inline src/components/pacbio/PacbioWellModal.vue]: Here too.

File: src/components/pacbio/PacbioWellModal.vue [hunk 5]
```diff
@@ -175,7 +190,7 @@ export default {
       this.mutateWell({ position: this.position, property: 'pre_extension_time', with: preExtensionTime })
     },
     updateGenerateHiFi(generateHiFi) {
-      ((generateHiFi !== "On Instrument" ) ? this.updateCCSAnalysisOutput("") : "") //sets CCSAnalysis back to blank if OnInstrument is unselected
+      // ((generateHiFi !== "On Instrument" ) ? this.updateCCSAnalysisOutput("") : "") //sets CCSAnalysis back to blank if OnInstrument is unselected
```
Comment (stevieing) [inline src/components/pacbio/PacbioWellModal.vue]: commented out.

File: src/components/pacbio/PacbioWellModal.vue [hunk 6]
```diff
@@ -211,9 +231,9 @@ export default {
     },
   },
   computed: {
-    showCCSAnalysis() {
-      return ((this.currentRun.system_name == "Sequel IIe" && this.generateHiFi == "On Instrument") ? true : false)
-    },
+    // showCCSAnalysis() {
```
Comment (stevieing) [inline src/components/pacbio/PacbioWellModal.vue]: commented out. Guess this is all to do with the merge conflicts.

File: tests/unit/components/pacbio/PacbioWellModal.spec.js [hunk 7]
```diff
@@ -32,6 +32,14 @@ describe('PacbioRunInfo', () => {
     expect(modal.movieTimeOptions).toEqual([{ text: 'Movie Time', value: "" }, "15.0", "20.0", "24.0", "30.0"])
   })
 
+  it('must have ccsAnalysisOutputOptions data', () => {
+    expect(modal.ccsAnalysisOutputOptions).toEqual(['Yes', 'No'])
+  })
+
+  // it('must have cssAnalysisOptions data', () => {
```
Comment (stevieing) [inline tests/unit/components/pacbio/PacbioWellModal.spec.js]: I think this might be the last one ...

**Review by stevieing — APPROVED** [review]
All good. Just noticed a few things that were commented out.

File: src/api/PacbioRun.js [hunk 8]
```diff
@@ -1,41 +1,19 @@
...
-
-const rows = () => {
-  return Array.from(Array(NO_OF_ROWS), (e, i) => String.fromCharCode(65 + i))
-}
-
-const buildWell = (row, column) => {
-  return  {
-    row: row, 
-    column: column, 
+// const buildWell = (row, column, generateHiFiDefault='', preExtenstionTimeDefault
```
Comment (stevieing) [inline src/api/PacbioRun.js:3]: Apparently all of my comments about your comments were outdated but don't know if this comment is also outdated? :)

**Pull request merged**